File: windows/Scanner/EscalationEvaluator.cs
using AIExposureScanner.Scanner.Models;

namespace AIExposureScanner.Scanner;

public sealed class EscalationEvaluator
{
    public IReadOnlyList<Finding> Evaluate(IReadOnlyList<Finding> findings, IReadOnlyList<EscalationRule> rules)
    {
        var result = findings.ToList();
        foreach (var rule in rules)
        {
            result = Apply(rule, result);
        }
        return result;
    }

    private static List<Finding> Apply(EscalationRule rule, List<Finding> findings) =>
        rule.Scope switch
        {
            EscalationScope.PerServer => ApplyPerServer(rule, findings),
            EscalationScope.Global    => ApplyGlobal(rule, findings),
            _                         => findings
        };

    private static List<Finding> ApplyPerServer(EscalationRule rule, List<Finding> findings)
    {
        var groups = findings
            .GroupBy(f => $"{f.App}||{f.ServerName ?? string.Empty}", StringComparer.Ordinal)
            .ToList();

        var result = findings.ToList();
        foreach (var group in groups)
        {
            var ruleIds = group.Select(f => f.RuleId).ToHashSet(StringComparer.Ordinal);
            if (!rule.RequiresRuleIds.IsSubsetOf(ruleIds)) continue;

            result = result
                .Select(f =>
                    group.Contains(f) &&
                    rule.RequiresRuleIds.Contains(f.RuleId) &&
                    (int)f.Severity > (int)rule.EscalateTo
                        ? f with { Severity = rule.EscalateTo, EscalationReason = rule.Reason }
                        : f)
                .ToList();
        }
        return result;
    }

    private static List<Finding> ApplyGlobal(EscalationRule rule, List<Finding> findings)
    {
        var allRuleIds = findings.Select(f => f.RuleId).ToHashSet(StringComparer.Ordinal);
        if (!rule.RequiresRuleIds.IsSubsetOf(allRuleIds)) return findings;

        if (rule.MinimumDistinctServers > 1)
        {
            foreach (var reqId in rule.RequiresRuleIds)
            {
                var distinctServers = findings
                    .Where(f => f.RuleId == reqId && f.ServerName is not null)
                    .Select(f => f.ServerName!)
                    .Distinct(StringComparer.Ordinal)
                    .Count();
                if (distinctServers < rule.MinimumDistinctServers) return findings;
            }
        }

        return findings
            .Select(f =>
                rule.RequiresRuleIds.Contains(f.RuleId) && (int)f.Severity > (int)rule.EscalateTo
                    ? f with { Severity = rule.EscalateTo, EscalationReason = rule.Reason }
                    : f)
            .ToList();
    }
}

File: macos/Scanner/Sources/Scanner/EscalationEvaluator.swift
public struct EscalationEvaluator: Sendable {
    public init() {}

    public func evaluate(_ findings: [Finding], rules: [EscalationRule]) -> [Finding] {
        var result = findings
        for rule in rules {
            result = apply(rule, to: result)
        }
        return result
    }

    private func apply(_ rule: EscalationRule, to findings: [Finding]) -> [Finding] {
        switch rule.scope {
        case .perServer:
            return applyPerServer(rule, to: findings)
        case .global:
            return applyGlobal(rule, to: findings)
        }
    }

    private func applyPerServer(_ rule: EscalationRule, to findings: [Finding]) -> [Finding] {
        let groups = Dictionary(grouping: findings) { "\($0.app)||\($0.serverName ?? "")" }
        var result = findings
        for (_, group) in groups {
            let ruleIds = Set(group.map(\.ruleId))
            guard rule.requiresRuleIds.isSubset(of: ruleIds) else { continue }
            result = result.map { finding in
                guard
                    group.contains(finding),
                    rule.requiresRuleIds.contains(finding.ruleId),
                    finding.severity.sortRank > rule.escalateTo.sortRank
                else { return finding }
                return escalated(finding, to: rule)
            }
        }
        return result
    }

    private func applyGlobal(_ rule: EscalationRule, to findings: [Finding]) -> [Finding] {
        let allRuleIds = Set(findings.map(\.ruleId))
        guard rule.requiresRuleIds.isSubset(of: allRuleIds) else { return findings }

        if rule.minimumDistinctServers > 1 {
            for reqId in rule.requiresRuleIds {
                let distinctServers = Set(findings.filter { $0.ruleId == reqId }.compactMap(\.serverName))
                if distinctServers.count < rule.minimumDistinctServers { return findings }
            }
        }

        return findings.map { finding in
            guard
                rule.requiresRuleIds.contains(finding.ruleId),
                finding.severity.sortRank > rule.escalateTo.sortRank
            else { return finding }
            return escalated(finding, to: rule)
        }
    }

    private func escalated(_ finding: Finding, to rule: EscalationRule) -> Finding {
        Finding(
            ruleId: finding.ruleId,
            severity: rule.escalateTo,
            app: finding.app,
            serverName: finding.serverName,
            extensionId: finding.extensionId,
            affectedPath: finding.affectedPath,
            maskedValue: finding.maskedValue,
            title: finding.title,
            explanation: finding.explanation,
            recommendation: finding.recommendation,
            escalationReason: rule.reason
        )
    }
}
